### PR TITLE
Add build-time flag to control number of instructions

### PR DESCRIPTION
**Pull request by seagater**
This PR depends on the previous PR:
Merge nccl-2.23.4-1 to msccl-executor-nccl https://github.com/Azure/msccl-executor-nccl/pull/47

The latest commit is to add build-time flag MSCCL_MAX_NUM_STEPS to control the number of instructions. Default value is 64.

### CODE REVIEW

**Comment by yzygitzh** [comment]
Looks good to me, please help rebase this against latest main, thanks!

**Review by yzygitzh — APPROVED** [review]

**Comment by yzygitzh** [comment]
How shall we set the MSCCL_MAX_NUM_STEPS from outside? I tried
`MSCCL_MAX_NUM_STEPS=64 make -j src.build NVCC_GENCODE="-gencode=arch=compute_80,code=sm_80"`
and
`make -j src.build NVCC_GENCODE="-gencode=arch=compute_80,code=sm_80 MSCCL_MAX_NUM_STEPS=64 "
and both of them won't work.

**Comment by yzygitzh** [comment]
Btw, instead of root `Makefile` I think we'd better revise `makefiles/common.mk` where seems most options are placed.

**Comment by seagater** [comment]
> How shall we set the MSCCL_MAX_NUM_STEPS from outside? I tried revising original MSCCL_MAX_NUM_STEPS to some invalid string and used `MSCCL_MAX_NUM_STEPS=64 make -j src.build NVCC_GENCODE="-gencode=arch=compute_80,code=sm_80"` and `make -j src.build NVCC_GENCODE="-gencode=arch=compute_80,code=sm_80" MSCCL_MAX_NUM_STEPS=64` but both of them won't work.

Do you mean the build failure for this flag? I tried the second usage and can build the library.

**Comment by seagater** [comment]
> Btw, instead of root `Makefile` I think we'd better revise `makefiles/common.mk` where seems most options are placed.

Good idea. I'll move the changes to common.mk file.

**Comment by seagater** [comment]
@yzygitzh The issue of undefined macro has been fixed. Now the build flag can be passed to the head files. Also handle the case for head file included for .cu files.

File: Makefile [hunk 1]
```diff
@@ -27,5 +27,6 @@ src.%:
 pkg.%:
 	${MAKE} -C pkg $* BUILDDIR=${ABSBUILDDIR}
 
+
```
Comment (yzygitzh) [inline Makefile:30]: Redundant whitespace

File: src/include/msccl/msccl_struct.h [hunk 2]
```diff
@@ -14,7 +14,9 @@
 #include "device.h"
 #include "msccl/msccl_scheduler.h"
 
+#ifndef MSCCL_MAX_NUM_STEPS
```
Comment (yzygitzh) [inline src/include/msccl/msccl_struct.h:17]: Shall we just remove `MSCCL_MAX_NUM_STEPS` here since it's duplicated macro definition now

**Review by yzygitzh — CHANGES_REQUESTED** [review]

File: src/include/msccl/msccl_struct.h (hunk 2 above)
Comment (seagater) [reply src/include/msccl/msccl_struct.h]: Agree. Now MSCCL_MAX_NUM_STEPS is removed from msccl_struct.h

File: Makefile (hunk 1 above)
Comment (seagater) [reply Makefile]: Resolved.

**Review by yzygitzh — APPROVED** [review]

**Pull request merged**